src/WidgetBase.ts

```diff
@@ -318,28 +318,48 @@ export class WidgetBase<P = WidgetProperties, C extends DNode = DNode> extends E
 
 	public __setProperties__(originalProperties: this['properties']): void {
 		const properties = this._runBeforeProperties(originalProperties);
-		const changedPropertyKeys: string[] = [];
-		const allProperties = [ ...Object.keys(properties), ...Object.keys(this._properties) ];
-		const checkedProperties: string[] = [];
-		const diffPropertyResults: any = {};
+		const alwaysRender = this.getDecorator('alwaysRender').length > 0;
 		const registeredDiffPropertyNames = this.getDecorator('registeredDiffProperty');
-		let runReactions = false;
 
+		let changedPropertyKeys: string[] = [];
+		let diffPropertyResults: any = {};
 		this._renderState = WidgetRenderState.PROPERTIES;
 
-		for (let i = 0; i < allProperties.length; i++) {
-			const propertyName = allProperties[i];
-			if (checkedProperties.indexOf(propertyName) > 0) {
-				continue;
+		if (alwaysRender) {
+			if (registeredDiffPropertyNames.length > 0) {
+				throw new Error('Widget marked as `alwayRender` but has custom diffs registered');
 			}
-			checkedProperties.push(propertyName);
-			const previousProperty = this._properties[propertyName];
-			const newProperty = this._bindFunctionProperty(properties[propertyName], this._coreProperties.bind);
-			if (registeredDiffPropertyNames.indexOf(propertyName) !== -1) {
-				runReactions = true;
-				const diffFunctions = this.getDecorator(`diffProperty:${propertyName}`);
-				for (let i = 0; i < diffFunctions.length; i++) {
-					const result = diffFunctions[i](previousProperty, newProperty);
+			changedPropertyKeys = Object.keys(properties);
+			diffPropertyResults = { ...properties };
+		}
+		else {
```

Comment:
Just curious... This seems to be far more complex logic for the "not always render case".  Is this realising when we address this logic uniquily, we have a far more nuanced invalidation mechanism that we need to employ?

Reply:
Actually it isn't specifically new LOC, it allows some sort of colocation of the code... Side note, this bit does feel a bit hefty from a complexity bit, I don't see any obvious ways of refactoring it, but just a thought...

Reply:
Indeed it's the existing diff'ing logic that was moved into the `else` of the `alwaysRender` `if`check. It is pretty complicated though, as every property can have a different diff registered, return a diff value to the passed property and also have reactions to run.